driver-core/src/main/com/mongodb/internal/connection/tlschannel/impl/BufferHolder.java

```diff
@@ -13,150 +13,149 @@
-        ByteBuffer newBuffer = newByteBuf.asNIO();
-        ((Buffer) buffer).flip();
-        newBuffer.put(buffer);
-        if (plainData) {
-            zero();
-        }
-        allocator.free(byteBuf);
-        byteBuf = newByteBuf;
-        buffer = newBuffer;
-        lastSize = newCapacity;
+  }
+
+  public void resize(int newCapacity) {
+    if (newCapacity > maxSize)
+      throw new IllegalArgumentException(
+          String.format(
+              "new capacity (%s) bigger than absolute max size (%s)", newCapacity, maxSize));
+    logger.trace(
+        "resizing buffer {}, increasing from {} to {} (manual sizing)",
+        name,
+        buffer.capacity(),
+        newCapacity);
+    resizeImpl(newCapacity);
+  }
+
+  public void enlarge() {
+    if (buffer.capacity() >= maxSize) {
+      throw new IllegalStateException(
+          String.format(
+              "%s buffer insufficient despite having capacity of %d", name, buffer.capacity()));
     }
-
-    /**
-     * Fill with zeros the remaining of the supplied buffer. This method does
-     * not change the buffer position.
-     * <p>
-     * Typically used for security reasons, with buffers that contains
-     * now-unused plaintext.
-     */
-    public void zeroRemaining() {
-        ((Buffer) buffer).mark();
-        buffer.put(ZEROS, 0, buffer.remaining());
-        ((Buffer) buffer).reset();
+    int newCapacity = Math.min(buffer.capacity() * 2, maxSize);
+    logger.trace(
+        "enlarging buffer {}, increasing from {} to {} (automatic enlarge)",
+        name,
+        buffer.capacity(),
+        newCapacity);
+    resizeImpl(newCapacity);
+  }
+
+  private void resizeImpl(int newCapacity) {
+    ByteBuffer newBuffer = allocator.allocate(newCapacity);
+    ((Buffer) buffer).flip();
+    newBuffer.put(buffer);
+    if (plainData) {
+      zero();
     }
-
-    /**
-     * Fill the buffer with zeros. This method does not change the buffer position.
-     * <p>
-     * Typically used for security reasons, with buffers that contains
-     * now-unused plaintext.
-     */
-    public void zero() {
-        ((Buffer) buffer).mark();
-        ((Buffer) buffer).position(0);
-        buffer.put(ZEROS, 0, buffer.remaining());
-        ((Buffer) buffer).reset();
+    allocator.free(buffer);
+    buffer = newBuffer;
+    lastSize = newCapacity;
+  }
+
+  /**
+   * Fill with zeros the remaining of the supplied buffer. This method does not change the buffer
+   * position.
+   *
+   * <p>Typically used for security reasons, with buffers that contains now-unused plaintext.
+   */
+  public void zeroRemaining() {
+    zero(buffer.position());
+  }
+
+  /**
+   * Fill the buffer with zeros. This method does not change the buffer position.
+   *
+   * <p>Typically used for security reasons, with buffers that contains now-unused plaintext.
+   */
+  public void zero() {
+    zero(0);
+  }
+
+  private void zero(final int position) {
```

Comment:
I'm not clear why this is necessary. `PowerOfTwoBufferPool` will set the limit to whatever size is requested by the caller, so as long as that size is less than or equal to `TlsChannelImpl.maxTlsPacketSize` we shouldn't have to loop.

Reply:
There are various calls to` buffer.compact`that resets the limit to the capacity of the bytebuffer

Reply:
Ah, that's unfortunate.